lib/charms/hpc_libs/v0/slurm_ops.py

```diff
@@ -674,6 +634,28 @@ def _create_state_save_location(self) -> None:
     def _apply_overrides(self) -> None:
         """Override defaults supplied provided by Slurm Debian packages."""
         match self._service_name:
+            case "sackd":
+                _logger.debug("overriding default sackd service configuration")
+                config_override = Path(
+                    "/etc/systemd/system/sackd.service.d/10-sackd-config-server.conf"
+                )
+                config_override.mkdir(parents=True, exist_ok=True)
+                config_override.write_text(
+                    textwrap.dedent(
+                        """
+                        [Service]
+                        ExecStart=
+                        ExecStart=/usr/bin/sh -c "/usr/sbin/sackd --systemd $${SACKD_CONFIG_SERVER:+--conf-server $$SACKD_CONFIG_SERVER} $$SACKD_OPTIONS"
+                        """
```

Comment:
Could we make this:

```
"""
[Service]
ExecStart=
ExecStart=/usr/bin/sh -c "/usr/sbin/sackd $${SACKD_CONFIG_SERVER:+--conf-server $$SACKD_CONFIG_SERVER} $$SACKD_OPTIONS"
Type=forking
"""
```

At the moment, sackd.service fails to start with:

```
systemd[1]: sackd.service: Got notification message from PID 7809, but reception only permitted for main PID 7808
systemd[1]: sackd.service: start operation timed out. Terminating.
```

Adding `Type=forking` gets you to `sh[7101]: sackd: fatal: Missing NOTIFY_SOCKET` then removing `--systemd` allows the service to start.

Alternatively, setting `SACKD_OPTIONS="--conf-server <hostname>:6817"` and having no config override could be an idea (e.g. through `config_server`?)

Reply:
Hmm... not sure we should use Type=forking here. Mostly because it's discouraged by systemd: https://www.freedesktop.org/software/systemd/man/latest/systemd.service.html#Options

I think we can do some magic in the sackd environment file to get around needing to start sackd up in a child process that needs to be monitored.

Reply:
Agreed with the environment file. Getting the config into `SACKD_OPTIONS` in `/etc/default/sackd` means we wouldn't need any override and could remove the `case "sackd":` branch of `_apply_overrides()` entirely.

Similarly, the `slurmd` case could have its options moved to `SLURMD_OPTIONS` in `/etc/default/sackd`, and [its use of `Type=forking` reassessed](https://github.com/charmed-hpc/slurm-charms/blob/main/charms/slurmd/src/utils/service.py#L55), I say cautiously, not wanting to open the `service.py` can of worms...

Reply:
I got it such that we don't need to use `Type=forking` for the daemon. Rather than perform the variable substitution directly in the service file, we instead do it the _/etc/default/sackd_ file used by the systemd service file.

```shell
SACKD_CONFIG_SERVER=''
SACKD_OPTIONS="${SACKD_CONFIG_OPTIONS:+--conf-server $SACKD_CONFIG_SERVER}"
``` 

Key thing to note is that `SACKD_CONFIG_SERVER` will always need to come before `SACKD_OPTIONS`, otherwise the variable substitution will not be performed correctly when sourcing in the environment file.

Reply:
Hmm, this isn't working in my testing. The `sackd` service fails to start when _/etc/default/sackd_ is:

```
SACKD_CONFIG_SERVER='juju-46b4c3-0:6817'
SACKD_OPTIONS="${SACKD_CONFIG_SERVER:+--conf-server $SACKD_CONFIG_SERVER}"
```

but starts if I manually edit it to:

```
SACKD_OPTIONS="--conf-server juju-46b4c3-0:6817"
```

Details below:

```
ubuntu@juju-46b4c3-1:~$ sudo systemctl restart sackd
Job for sackd.service failed because the control process exited with error code.
See "systemctl status sackd.service" and "journalctl -xeu sackd.service" for details.
ubuntu@juju-46b4c3-1:~$ sudo systemctl status sackd
× sackd.service - Slurm auth and cred kiosk daemon
     Loaded: loaded (/usr/lib/systemd/system/sackd.service; enabled; preset: enabled)
     Active: failed (Result: exit-code) since Fri 2024-12-06 18:03:07 UTC; 1s ago
       Docs: man:sackd(8)
    Process: 5262 ExecStart=/usr/sbin/sackd --systemd $SACKD_OPTIONS (code=exited, status=1/FAILURE)
   Main PID: 5262 (code=exited, status=1/FAILURE)
        CPU: 6ms

Dec 06 18:03:07 juju-46b4c3-1 systemd[1]: Starting sackd.service - Slurm auth and cred kiosk daemon...
Dec 06 18:03:07 juju-46b4c3-1 sackd[5262]: sackd: error: resolve_ctls_from_dns_srv: res_nsearch error: Unknown host
Dec 06 18:03:07 juju-46b4c3-1 sackd[5262]: sackd: error: fetch_config: DNS SRV lookup failed
Dec 06 18:03:07 juju-46b4c3-1 sackd[5262]: sackd: fatal: _establish_config_source: failed to load configs
Dec 06 18:03:07 juju-46b4c3-1 systemd[1]: sackd.service: Main process exited, code=exited, status=1/FAILURE
Dec 06 18:03:07 juju-46b4c3-1 systemd[1]: sackd.service: Failed with result 'exit-code'.
Dec 06 18:03:07 juju-46b4c3-1 systemd[1]: Failed to start sackd.service - Slurm auth and cred kiosk daemon.
ubuntu@juju-46b4c3-1:~$ sudo cat /etc/default/sackd 
SACKD_CONFIG_SERVER='juju-46b4c3-0:6817'
SACKD_OPTIONS="${SACKD_CONFIG_SERVER:+--conf-server $SACKD_CONFIG_SERVER}"
ubuntu@juju-46b4c3-1:~$ sudo nano /etc/default/sackd 
ubuntu@juju-46b4c3-1:~$ sudo cat /etc/default/sackd 
SACKD_OPTIONS="--conf-server juju-46b4c3-0:6817"
ubuntu@juju-46b4c3-1:~$ sudo systemctl restart sackd
ubuntu@juju-46b4c3-1:~$ sudo systemctl status sackd
● sackd.service - Slurm auth and cred kiosk daemon
     Loaded: loaded (/usr/lib/systemd/system/sackd.service; enabled; preset: enabled)
     Active: active (running) since Fri 2024-12-06 18:04:28 UTC; 3s ago
       Docs: man:sackd(8)
   Main PID: 5323 (sackd)
      Tasks: 11
     Memory: 1.6M (peak: 2.0M)
        CPU: 9ms
     CGroup: /system.slice/sackd.service
             └─5323 /usr/sbin/sackd --systemd --conf-server juju-46b4c3-0:6817

Dec 06 18:04:28 juju-46b4c3-1 systemd[1]: Starting sackd.service - Slurm auth and cred kiosk daemon...
Dec 06 18:04:28 juju-46b4c3-1 sackd[5323]: sackd: running
Dec 06 18:04:28 juju-46b4c3-1 systemd[1]: Started sackd.service - Slurm auth and cred kiosk daemon.
ubuntu@juju-46b4c3-1:~$
```

Reply:
In the new commit https://github.com/charmed-hpc/hpc-libs/pull/55/commits/1b967905297d813a2dd0731bd2b153054b13286d, @jedel1043 and I decided that we should just mandate that the configuration server is provided rather than try to be clever with shell variable substitution. This way we can ensure that `sackd` is started the way we want it started, and we don't need to make `sackd` `Type=forking`.

Here's new implementation working in a my development playground.

```
● sackd.service - Slurm auth and cred kiosk daemon
     Loaded: loaded (/usr/lib/systemd/system/sackd.service; enabled; preset: enabled)
    Drop-In: /etc/systemd/system/sackd.service.d
             └─10-sackd-config-server.conf
     Active: active (running) since Fri 2024-12-06 20:21:39 UTC; 1min 4s ago
       Docs: man:sackd(8)
   Main PID: 2753 (sackd)
      Tasks: 11
     Memory: 900.0K (peak: 2.5M)
        CPU: 9ms
     CGroup: /system.slice/sackd.service
             └─2753 /usr/sbin/sackd --systemd --conf-server sackd-playground:6817

Dec 06 20:21:39 sackd-playground systemd[1]: Starting sackd.service - Slurm auth and cred kiosk daemon...
Dec 06 20:21:39 sackd-playground sackd[2753]: sackd: warning: sackd running as slurm instead of SlurmUser(root)
Dec 06 20:21:39 sackd-playground sackd[2753]: sackd: running
Dec 06 20:21:39 sackd-playground systemd[1]: Started sackd.service - Slurm auth and cred kiosk daemon.
```